### PR TITLE
Unsloth-AutoGPTQ Plugin

### PR DESCRIPTION
**Pull request by achew010**
### Description

Things to document
- ~~license issue~~
- ~~only supports dropout == 0~~
- has a dependency on optimum that we should get rid off. Also has a dependeny on xformers.
- ~~seems that quantization config not properly propagated down into `FastLanguageModel`~~
- document that we need to use fp16

This PR introduces a new plugin to `tuning/acceleration/plugins` that provides the following triton optimizations in addition to the AutoGPTQ TritonV2 base layer.

1. Loss
2. LayerNorm
3. RoPE Embeddings
4. Fused Attention

#### NOTE:

1. Unsloth requires lora_dropout to be set to 0 for fast patching. if lora_config.dropout > 0, unsloth will skip replacing attention modules with fast modules.
2. Unsloth's OSS license allows for up to 4 GPUs, will require a commercial license for larger resources https://github.com/unslothai/unsloth/blob/d215fd902cf28feb8abcfde2d25281d0fbf9d28c/unsloth/models/llama.py#L1140-L1143

#### 1. Unsloth 
Unsloth has a dependency on the xformers library that will require a specific tag corresponding to the user's installation environment (cuda version, pytorch version etc.) 
Refer to Installation Instructions in [Unsloth](https://github.com/unslothai/unsloth) `README.md` for more information
  
EXAMPLE: `pip install git+https://$GHE_REPO/unsloth.git@dev .[cu121_ampere_torch220]`

### Reproducibility
```
python \
    tuning/sft_trainer.py \
    --training_data_path $DATA_PATH \
    --model_name_or_path TheBloke/Mixtral-8x7B-Instruct-v0.1-GPTQ \
    --acceleration_framework_config_file fixtures/acceleration_framework.yaml \
    --output_dir ./results \
    --num_train_epochs 1 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --gradient_checkpointing True \
    --torch_dtype float16 \
    --fp16 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0. \
    --lr_scheduler_type "linear" \
    --logging_strategy 'steps' \
    --logging_steps 10 \
    --include_tokens_per_second \
    --packing True \
    --use_flash_attn False \
    --response_template "\n### Response:" \
    --dataset_text_field "output" \
    --peft_method lora \
    --evaluation_strategy "no" \
    --save_strategy "no" \
    --max_steps 200 \
    --target_modules q_proj k_proj v_proj o_proj \
    --r 16 \
    --lora_dropout 0 \
    --lora_alpha 16
```

### Tests

A test was carried out to show the similarity of throughput and loss between the new plugin to FMS and the custom unsloth script. 
* Optimized Loss, LayerNorm and RoPE embeddings
* Fused LoRA Operations with Q, K, V and O in the Attention Module
* Note that there are no fused LoRA operations in the MOE blocks of Mixtral in the following examples.

### CODE REVIEW

File: fixtures/acceleration_framework.yaml [hunk 1]
```diff
@@ -5,11 +5,14 @@ plugins:
 
     # quantization
     quantization: 
+      # Include unsloth optimizations
+      unsloth_autogptq:
```
Comment (fabianlim) [inline fixtures/acceleration_framework.yaml]: lets have do the config this way

```yaml
quantization:
    unsloth:
        base_layer:  auto_gptq
        kernel: triton_v2
```

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        self, 
+        model, 
+        train_args: TrainingArguments,
+        modifiable_args: Tuple[LoraConfig],
+    ):
+        peft_config, = modifiable_args # unpack modifiable args
+
+        # some assertions
+        assert peft_config is not None, "need peft_config to install PEFT adapters"
+        assert model.dtype == torch.float16 or train_args.fp16,\
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py:101]: assert the lora_dropout here, just like we did when checking `peft_config is None`

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        # 1. investigate using BaseGPTQForCausalLM.make_sure_compatible_with_peft
+        #    to see if we can get around the patching
+
+        _old_create_new_module = LoraModel
+        _old_replace_module = GPTQLoraModel._replace_module
+        LoraModel._create_new_module = staticmethod(_create_new_module_triton)
+        GPTQLoraModel._replace_module = MethodType(_replace_module, GPTQLoraModel)
+
+        # In the unsloth implementation, the prepare_model_for_kbit_training get_peft_model is called 
+        # inside FastLanguageModel.patch_peft_model
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py]: if this is called inside `get_peft_model` it has to be explicitly called out

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    if (
+        isinstance(target, GPTQuantLinear)
+    ):
+        new_module = LoraLinearGPTQ(target, adapter_name, **kwargs)
+    # if module cannot be found, return None which results in a raise in the call-stack
+    return new_module
+
+class UnslothAutoGPTQAccelerationPlugin(AccelerationPlugin):
+    
+    require_packages = ['auto_gptq', 'unsloth']
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py]: need to add a required model archs here since we are limited to only few models (e.g. llama, mistral, mixtarl)

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        peft_config, = modifiable_args # unpack modifiable args
+
+        # some assertions
+        assert peft_config is not None, "need peft_config to install PEFT adapters"
+        assert model.dtype == torch.float16 or train_args.fp16,\
+             "need to run in fp16 mixed precision or load model in fp16"
+        peft_kwargs = peft_config.to_dict()
+        assert peft_kwargs['lora_dropout'] == 0, "Unsloth Fused Attention requires lora_dropout argument to be set to 0"
+
+        peft_kwargs.pop('task_type')
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py]: to improve readiblity, i wont convert and pop the 'task type'

1. check peft_config.lora_dropout == 0
2. then when passing kwargs do
```python
func(**{k:v for k,v in peft_config.to_dict() if k != 'task_type'})
```

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    def model_loader(self, model_name: str, **kwargs):
+        # 1. Load the gptq base model through unsloth FastLanguageModel
+        model, _ = FastLanguageModel.from_pretrained(
+            model_name,
+            quantization_method=QuantizationMethod.GPTQ,
+            device_map="auto",
+        )
+
+        # 2. Depending on the model, replace appropriate CUDA kernel with TritonV2 kernel 
+        if model.config.model_type == 'mixtral':
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py:61]: I remembner you had an if else staetment before to either load mixtral or others, can we have the old logic back?

This logic is confusing, because its patch style, where you need to know what the default `QuantLinear` is loaded by `FastLanguageModel.from_pretrained` toi understand why you need an if-else statement

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    def model_loader(self, model_name: str, **kwargs):
+        # 1. Load the gptq base model through unsloth FastLanguageModel
+        model, _ = FastLanguageModel.from_pretrained(
+            model_name,
+            quantization_method=QuantizationMethod.GPTQ,
+            device_map="auto",
+        )
+
+        # 2. Depending on the model, replace appropriate CUDA kernel with TritonV2 kernel
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py:64]:
```suggestion
        # 2. Depending on the model, replace appropriate CUDA kernel with TritonV2 kernel 
        # - `FastLanguageModel` above will load a "default" cuda linear kernel. 
        # - the idea is to load the appropriate "default" kernel (for each model), then follow up with an injection
        #.  via the `inject_to_model(..., target_module_type=DefaultKernel)` method.
```

Comment (fabianlim) [reply tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py:64]: improved the comments for now. This logic is not very satisfying, but ok for now.

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
+
+def parse_configuration():
+    pass
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py]:
```suggestion
```
we do not need this.

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+class UnslothAutoGPTQAccelerationPlugin(AccelerationPlugin):
+    
+    require_packages = ['auto_gptq', 'unsloth', 'optimum']
+    restricted_model_archs = ['MixtralForCausalLM', 'LlamaForCausalLM', 'MistralForCausalLM', 'GemmaForCausalLM']
+
+    '''
+    NOTE:
+    Unsloth's OSS license allows for up to 4 GPUs, will require a commercial license for larger resources
+    https://github.com/unslothai/unsloth/blob/d215fd902cf28feb8abcfde2d25281d0fbf9d28c/unsloth/models/llama.py#L1140-L1143
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py]:
```suggestion
    NOTE: unloth's LICENSE file looks like a standar Apache 2.0, but in various parts of the code, it claims to require a commercial license if used to run on more than 4 GPUs, see 
    https://github.com/unslothai/unsloth/blob/d215fd902cf28feb8abcfde2d25281d0fbf9d28c/unsloth/models/llama.py#L1140-L1143
```

**Review by fabianlim — CHANGES_REQUESTED** [review]

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        # just do checking, nothing must to configure at this point
+        # if need to configure then do something like this:
+        # self.kernel = self._get_config_value("peft.kernel")
+        self._check_config_equal(key="peft.quantization", value="unsloth")
+        self._check_config_equal(key="peft.quantization.unsloth.base_layer", value="auto_gptq")
+        self._check_config_equal(key="peft.quantization.unsloth.kernel", value="triton_v2")
+
+    def model_loader(self, model_name: str, **kwargs):
+        # 1. Load the gptq base model through unsloth FastLanguageModel
+        torch_dtype = kwargs.get('torch_dtype', torch.float16)
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py]:
```suggestion
        torch_dtype = kwargs.get('torch_dtype', torch.float32)
```
For consistency

File: tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        self._check_config_equal(key="peft.quantization.unsloth.base_layer", value="auto_gptq")
+        self._check_config_equal(key="peft.quantization.unsloth.kernel", value="triton_v2")
+
+    def model_loader(self, model_name: str, **kwargs):
+        # 1. Load the gptq base model through unsloth FastLanguageModel
+        torch_dtype = kwargs.get('torch_dtype', torch.float32)
+        model, _ = FastLanguageModel.from_pretrained(
+                model_name,
+                dtype=torch_dtype,
+                device_map="auto",
```
Comment (fabianlim) [inline tuning/acceleration/plugins/framework_plugin_unsloth_autogptq.py]: do we really need `device_map`='auto' here?
